### PR TITLE
[serve] Add `ReplicaID` class

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

This is stacked on: https://github.com/ray-project/ray/pull/43555

Unifies replica IDs on a `ReplicaID` class (similar to and including `DeploymentID`). This is used to uniquely identify replicas and carry around their metadata.

This allows us to clean up the strings in some of our logs (where the replica_id contained redundant information about the application and deployment name).

Example logs from a "hello world" example:
```
INFO 2024-02-29 10:30:55,102 controller 85325 controller.py:403 - Finished recovering deployments after 0.00s.
INFO 2024-02-29 10:30:55,103 controller 85325 proxy_state.py:150 - Starting proxy with name 'SERVE_PROXY_ACTOR-bad9879d63539f90c4bc47f6b31d4fa210ee88a5164de09a65bb38a6' on node 'bad9879d63539f90c4bc47f6b31d4fa210ee88a5164de09a65bb38a6' listening on '127.0.0.1:8000'
INFO 2024-02-29 10:30:55,926 controller 85325 deployment_state.py:1578 - Deploying new version of deployment A in application 'default'. Setting initial target number of replicas to 1.
INFO 2024-02-29 10:30:56,029 controller 85325 deployment_state.py:1883 - Adding 1 replica to deployment A in application 'default'.
INFO 2024-02-29 10:30:56,029 controller 85325 deployment_state.py:393 - Starting Replica(id='xujwc8b6', deployment='A', app='default').
INFO 2024-02-29 10:30:56,860 controller 85325 deployment_state.py:2022 - Replica(id='xujwc8b6', deployment='A', app='default') started successfully on node bad9879d63539f90c4bc47f6b31d4fa210ee88a5164de09a65bb38a6.
```

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by edoakes** [comment]
Suggest to wait until other PR is merged before reviewing.

**Comment by edoakes** [comment]
Bad merge, randomly tagged people please ignore

**Comment by edoakes** [comment]
Ready for review

File: python/ray/serve/_private/common.py
Comment (GeneDer) [inline python/ray/serve/_private/common.py:44]: Non-blocker: Can probably just  rename this to something like `full_id` and make it a property

**Review by GeneDer — APPROVED** [review]
LGTM!

File: python/ray/serve/_private/common.py
Comment (edoakes) [reply python/ray/serve/_private/common.py:90]: roger. forgot to update :)

**Pull request merged**
